File: android/ProximitySensor.m
#import "ProximitySensor.h"
#import <UIKit/UIKit.h>

@implementation ProximitySensor {
    BOOL _hasListeners;
}

RCT_EXPORT_MODULE();

- (NSArray<NSString *> *)supportedEvents {
    return @[@"proximityChange", @"proximityError"];
}

- (void)startObserving {
    _hasListeners = YES;
    [[UIDevice currentDevice] setProximityMonitoringEnabled:YES];
    [[NSNotificationCenter defaultCenter] addObserver:self
                                           selector:@selector(proximityStateDidChange:)
                                               name:UIDeviceProximityStateDidChangeNotification
                                             object:nil];
}

- (void)stopObserving {
    _hasListeners = NO;
    [[UIDevice currentDevice] setProximityMonitoringEnabled:NO];
    [[NSNotificationCenter defaultCenter] removeObserver:self
                                                  name:UIDeviceProximityStateDidChangeNotification
                                                object:nil];
}

- (void)proximityStateDidChange:(NSNotification *)notification {
    if (!_hasListeners) return;
    
    UIDevice *device = [notification object];
    NSDictionary *payload = @{
        @"isNear": @(device.proximityState),
        @"value": @(device.proximityState ? 0 : 1), // Simple binary value for iOS
        @"timestamp": @([NSDate date].timeIntervalSince1970 * 1000)
    };
    
    [self sendEventWithName:@"proximityChange" body:payload];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

RCT_EXPORT_METHOD(isAvailable:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    BOOL isAvailable = [UIDevice currentDevice].isProximityMonitoringEnabled;
    resolve(@(isAvailable));
}

RCT_EXPORT_METHOD(startProximitySensor:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
    dispatch_async(dispatch_get_main_queue(), ^{
        if (![UIDevice currentDevice].isProximityMonitoringEnabled) {
            [[UIDevice currentDevice] setProximityMonitoringEnabled:YES];
        }
        
        if ([UIDevice currentDevice].isProximityMonitoringEnabled) {
            resolve(@YES);
        } else {
            reject(@"PROXIMITY_ERROR", @"Failed to enable proximity monitoring", nil);
        }
    });
}

RCT_EXPORT_METHOD(stopProximitySensor) {
    dispatch_async(dispatch_get_main_queue(), ^{
        [[UIDevice currentDevice] setProximityMonitoringEnabled:NO];
    });
}

+ (BOOL)requiresMainQueueSetup {
    return YES;
}

@end

File: android/ProximitySensor.swift
import Foundation
import UIKit

@objc(ProximitySensor)
class ProximitySensor: RCTEventEmitter {
    private var hasListeners = false
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    override func supportedEvents() -> [String]! {
        return ["proximityChange"]
    }
    
    override func startObserving() {
        hasListeners = true
        UIDevice.current.isProximityMonitoringEnabled = true
        NotificationCenter.default.addObserver(
            self,
            selector: #selector(proximityStateChanged),
            name: UIDevice.proximityStateDidChangeNotification,
            object: nil
        )
    }
    
    override func stopObserving() {
        hasListeners = false
        UIDevice.current.isProximityMonitoringEnabled = false
        NotificationCenter.default.removeObserver(
            self,
            name: UIDevice.proximityStateDidChangeNotification,
            object: nil
        )
    }
    
    @objc func proximityStateChanged(notification: NSNotification) {
        if hasListeners {
            let isNear = UIDevice.current.proximityState
            let timestamp = NSDate().timeIntervalSince1970 * 1000 // Convert to milliseconds
            
            sendEvent(
                withName: "proximityChange",
                body: [
                    "isNear": isNear,
                    "value": isNear ? 0.0 : 1.0, // iOS doesn't provide exact value, so we simulate it
                    "timestamp": timestamp
                ]
            )
        }
    }
    
    @objc func isAvailable(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        resolve(true) // Proximity sensor is available on all iOS devices that support it
    }
    
    @objc func startProximitySensor(_ resolve: @escaping RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
        DispatchQueue.main.async {
            UIDevice.current.isProximityMonitoringEnabled = true
            resolve(true)
        }
    }
    
    @objc func stopProximitySensor() {
        DispatchQueue.main.async {
            UIDevice.current.isProximityMonitoringEnabled = false
        }
    }
    
    @objc func addListener(_ eventName: String) {
        // Keep: Required for RN built-in EventEmitter
    }
    
    @objc func removeListeners(_ count: Int) {
        // NativeEventEmitter requires this method
        if count == 0 {
            stopProximitySensor()
        }
    }
    
    override func invalidate() {
        stopProximitySensor()
        super.invalidate()
    }
}